### ISSUE TITLE
GraphBuilder: number of paths is limited to 254 in current implementation.

### ISSUE THREAD

**Issue by cgroza** [eblerjana/pangenie#79]
Hi,

```
version: v3.0.1
Files and parameters used:
-e      3000000000
-k      31
-o      index/processed
-r      chm13v2.0.fa
-t      10
-v      merged.sorted.multi.vcf
Determine allele sequences ...
Read reference genome ...
Found 25 chromosome(s) from the reference file.
Read input VCF ...
terminate called after throwing an instance of 'std::runtime_error'
  what():  GraphBuilder: number of paths is limited to 254 in current implementation.
```
I am hitting this error, trying to genotype a VCF with nearly 1000 genomes.
Do I need to reduce the number of genomes?
Otherwise, how difficult would it be to change pangenie to support more genomes?

**Comment by eblerjana** [comment]
The current implementation does not support more than 254 paths, the reason being that the size of the variables storing path information is limited. It should not be very difficult for me to fix this, but this would increase the memory requirements of the program significantly and also the algorithm is likely too slow to handle larger panels. We are working on an algorithm to subsample paths from the input VCF to make PanGenie applicable to larger panels, but we don't have anything ready for this yet unfortunately.

**Comment by Han-Cao** [comment]
Hi @eblerjana ,

To use the current pangenie to genotype large reference panel, I am wondering if it is appropriate to split it into small batches and then combine the output. For example:
- Use the pipeline to generate the graph and biallelic VCF with all samples
- Randomely subset the graph VCF into small VCFs (e.g., 100 samples) without overlapping, then filter variants by AC>1 and AN > 0.8 in subset graph VCFs (i.e., `bcftools view -S subset.txt graph.vcf | bcftools view -i 'AC>1 & F_MISSING>0.8'`)
- Run pangenie separately, convert to biallelic VCFs, and combine the output VCFs

I think this way can at least genotype most of the variants identified in the large panel, although we need to find a method to fix the inconsistent genotyping results among different batches. May I know what is your suggestion on it?

Many thanks.

**Comment by eblerjana** [comment]
Hi Han-Cao,

this sounds good, actually this is exactly what PanGenie already does internally to speed up genotyping. However, doing it as you suggested (i.e. "externally") might help reducing memory, so I think it's a good approach for now. You'd have to check if this is sufficiently fast (I don't have an estimate for it right now).

Best,
Jana

**Comment by Han-Cao** [comment]
Hi Jana,

This is an update on the performance of batch genotyping.

The reference panel is split into ~200 haplotypes batches (autosome only). It takes less than 4 hour per batch for a ~30X WGS data when using 32 cores. The F1 scores (reported by Truvari) are ~0.89 and are very similar between different batches. The current results are pretty good to me. 

As you mentioned that PanGenie internally do similar thing to speed up, may I know how I should merge the genotypes among external batches? Can I determine the genotype of a bubble using the one with highest GQ among all batches?

```
time spent reading UniqueKmersMap from disk:    21.1701 sec
time spent counting kmers in reads (wallclock):         1692.16 sec
time spent pre-computing probabilities:         0.0113615 sec
time spent updating unique kmers:       4299.54 sec
time spent selecting paths:     0.00023387 sec
time spent genotyping (total):  121976 sec
time spent writing output VCF:  940.293 sec
total wallclock time PanGenie-genotype: 13515.5 sec
Max RSS:        69.4046 GB
```

Besides, the pipeline I used to split reference is:

```
bcftools view -r $autosome -S $sample_batch1 -Ou $graph_vcf | \
bcftools +fill-tags -Ou -- -t AC,AN,AF | \
bcftools view --min-ac 1 | \
python3 ${scripts}/filter-vcf.py 0.8 | \
bcftools view --trim-alt-alleles -o ${outpath}/pangenie_batch1.graph.vcf

PanGenie-index \
-v ${outpath}/pangenie_batch1.graph.vcf \
-r $ref \
-t $threads \
-o ${outpath}/pangenie_batch1.index/pangenie_batch1.index
```

Thank you very much!

Best,
Han

**Comment by eblerjana** [comment]
Internally, the genotype likelihoods of each batch are added and the resulting likelihoods are normalized to obtain final genotype likelihoods for each bubble. The output VCF contains the phred-scaled genotype likelihoods. So I think one option would be to convert these back to probabilities, and then you can combine them with other batches in the same way.

**Comment by Han-Cao** [comment]
Thanks for the suggestions. I have read the `genotypingresult.cpp` code to try to understand what PanGenie internally do to normalize likelihoods and assign genotype. But I am not familiar with C++. Could you kindly clarify if the below workflow aligns to what you suggested:

Considering a complicated case, for a multiallelic bubble with 3 variants A, B, C, after `bcftools view --trim-alt-alleles`, the same bubble in full reference panel and batch VCFs are:
- **full.vcf**: A, B, C
- **batch1.vcf**: A, B
- **batch2.vcf**: A, C

To merge likelihoods and genotypes:
1. retrieve the genotype likelihoods from the `GL` field in batch VCFs
2. set the likelihood of missing genotypes as 0 (e.g., A/C in batch1.vcf, A/B in batch2.vcf)
3. add all likelihoods of the same genotypes across all batches
4. normalized likelihoods by sum
5. get the likeliest GT, where GT is set as missing if:
        - No GT with likelihood > 0
        - More than 1 likeliest GT

Some pseudocode:
```
# Convert existing GL to likelihoods,
# where index = ((a2* (a2 + 1)) / 2) + a1
likelihood_batch1[(a1, a2)] = 10^(-batch1_gl[index] / 10) 
likelihood_batch2[(a1, a2)] = 10^(-batch2_gl[index] / 10) 

# Set likelihood of missing GT as 0
likelihood_batch1[missing_pairs] = 0
likelihood_batch2[missing_pairs] = 0

# Add different batches
likelihood_merge[(a1,a2)] = likelihood_batch1[(a1, a2)] + likelihood_batch2[(a1, a2)]

# Normalize
total = sum(likelihood_merge)
if (total > 0) {
    likelihood_merge = likelihood_merge / total
}

# Get genotypes
best_gt = likelihood_merge[likelihood_merge.index_of_max()]
if (best_gt.size() > 1 or best_gt == 0) {
    best_gt = "./."
}
```

**Comment by eblerjana** [comment]
Yes exactly, this is how it is implemented currently in PanGenie.

**Comment by VLoegler** [comment]
Hi, 

I am also interested in using Pangenie with more than 254 haplotypes. @Han-Cao could you successfully modify Pangenie so it handles more haplotypes ?

**Comment by eblerjana** [comment]
As I mentioned above, I'm currently working on a version supporting larger panels. It is still under development, but can be tested already. In case anyone is interested in checking it out, it currently is available in the `sampling-tests` branch.  

This is how to (currently) run this new version:

**Indexing**  (same as before):
`PanGenie-index -r <reference.fa> -v <variants.vcf> -o preprocessing ` 

**Genotyping**  (add` -x 15` here, otherwise same as before):
`PanGenie -f preprocessing -i <reads.fa> -j <nr threads jellyfish>  -t <nr of threads genotyping> -o output -x 15`   

Note that this is still very experimental and I'm actively working to improve it, so use at your own risk. My preliminary benchmarking results look promising already though. Any feedback / suggestions are of course welcome!

**Comment by VLoegler** [comment]
Thanks a lot! I'll give it a try

**Comment by eblerjana** [comment]
Note: the `sampling-tests` branch is now merged to master and I have made a new release of PanGenie. As I wrote in my previous comment, this version now supports larger panels. Therefore, I'm closing this issue now.
